client/statement-store/src/lib.rs

```diff
@@ -0,0 +1,1263 @@
+						};
+						if let Some((_channel, len)) = account_rec.by_priority.get(&key) {
+							would_free_size = *len;
+							evicted.insert(channel_record.hash);
+						}
+					}
+				}
+			}
+			// Check if we can evict enough lower priority statements to satisfy constraints
+			for (entry, (_, len)) in account_rec.by_priority.iter() {
+				if (account_rec.data_size - would_free_size + statement_len <= max_size) &&
+					account_rec.by_priority.len() + 1 - evicted.len() <= max_count
+				{
+					// Satisfied
+					break
+				}
+				if evicted.contains(&entry.hash) {
+					// Already accounted for above
+					continue
+				}
+				if entry.priority >= priority {
+					log::debug!(
+						target: LOG_TARGET,
+						"Ignored message due to constraints {:?} {} < {}",
+						HexDisplay::from(&hash),
+						priority,
+						entry.priority,
+					);
+					return MaybeInserted::Ignored
+				}
+				evicted.insert(entry.hash);
+				would_free_size += len;
+			}
+		}
+		// Now check global constraints as well.
+		for (entry, len) in self.by_global_priority.iter() {
+			if (self.total_size - would_free_size + statement_len <= self.max_size) &&
+				self.by_global_priority.len() + 1 - evicted.len() <= self.max_entries
+			{
+				// Satisfied
+				break
+			}
+			if evicted.contains(&entry.hash) {
+				// Already accounted for above
+				continue
+			}
+
+			if entry.priority >= priority {
+				log::debug!(
+					target: LOG_TARGET,
+					"Ignored message due global to constraints {:?} {} < {}",
+					HexDisplay::from(&hash),
+					priority,
+					entry.priority,
+				);
+				return MaybeInserted::Ignored
+			}
+			evicted.insert(entry.hash);
+			would_free_size += len;
+		}
+
+		for h in &evicted {
+			self.make_expired(h, current_time);
+		}
+		self.insert_new(hash, *account, priority, statement);
+		MaybeInserted::Inserted(evicted)
+	}
+}
+
+impl Store {
+	/// Create a new shared store instance. There should only be one per process.
+	pub fn new_shared<Block, Client>(
+		path: &std::path::Path,
+		client: Arc<Client>,
+		prometheus: Option<&PrometheusRegistry>,
+	) -> Result<Arc<Store>>
+	where
+		Block: BlockT,
+		Block::Hash: From<BlockHash>,
+		Client: ProvideRuntimeApi<Block>
+			+ HeaderBackend<Block>
+			+ sc_client_api::ExecutorProvider<Block>
+			+ Send
+			+ Sync
+			+ 'static,
+		Client::Api: ValidateStatement<Block>,
+	{
+		let store = Arc::new(Self::new(path, client.clone(), prometheus)?);
+		client.execution_extensions().register_statement_store(store.clone());
+		Ok(store)
+	}
+
+	/// Create a new instance.
+	fn new<Block, Client>(
+		path: &std::path::Path,
+		client: Arc<Client>,
+		prometheus: Option<&PrometheusRegistry>,
+	) -> Result<Store>
+	where
+		Block: BlockT,
+		Block::Hash: From<BlockHash>,
+		Client: ProvideRuntimeApi<Block> + HeaderBackend<Block> + Send + Sync + 'static,
+		Client::Api: ValidateStatement<Block>,
+	{
+		let mut path: std::path::PathBuf = path.into();
+		path.pop();
+		path.push("statement");
```

Comment:
That assumes that `path` is in some special format. This should not be done.

Reply:
Fixed